--- Drive.pm	2025-06-09 18:31:45.482659331 +0200
+++ Drive.pm	2025-06-09 19:00:29.962125989 +0200
@@ -275,6 +275,14 @@
 	optional => 1,
 });
 
+my %iothread_vq_mapping_fmt = ( iothread_vq_mapping => {
+    type => 'integer',
+    description => "Whether to use iothread-vq-mapping for this drive",
+    minimum => 2,
+    maximum => 16,
+    optional => 1,
+});
+
 my %product_fmt = (
     product => {
 	type => 'string',
@@ -442,6 +450,7 @@
 my $virtio_fmt = {
     %drivedesc_base,
     %iothread_fmt,
+    %iothread_vq_mapping_fmt,
     %readonly_fmt,
 };
 my $virtiodesc = {
@@ -537,6 +546,7 @@
 my $alldrive_fmt = {
     %drivedesc_base,
     %iothread_fmt,
+    %iothread_vq_mapping_fmt,
     %model_fmt,
     %product_fmt,
     %queues_fmt,
